awx/main/dispatch/worker/callback.py

```diff
@@ -164,38 +163,48 @@ def flush(self, force=False):
                     else:  # only calculate the seconds if the created time already has been set
                         metrics_total_job_event_processing_seconds += e.modified - e.created
                 metrics_duration_to_save = time.perf_counter()
+                saved_events = []
                 try:
                     cls.objects.bulk_create(events)
                     metrics_bulk_events_saved += len(events)
+                    saved_events = events
+                    self.buff[cls] = []
                 except Exception as exc:
-                    logger.warning(f'Error in events bulk_create, will try indiviually up to 5 errors, error {str(exc)}')
+                    # If the database is flaking, let ensure_connection throw a general exception
+                    # will be caught by the outer loop, which goes into a proper sleep and retry loop
+                    django_connection.ensure_connection()
+                    logger.warning(f'Error in events bulk_create, will try indiviually, error: {str(exc)}')
                     # if an exception occurs, we should re-attempt to save the
                     # events one-by-one, because something in the list is
                     # broken/stale
-                    consecutive_errors = 0
-                    events_saved = 0
                     metrics_events_batch_save_errors += 1
-                    for e in events:
+                    for e in events.copy():
                         try:
                             e.save()
-                            events_saved += 1
-                            consecutive_errors = 0
+                            metrics_singular_events_saved += 1
+                            events.remove(e)
```

Comment:
Since `events` is a list, `.remove` is going to be doing a linear scan each loop iteration.  I'm not thrilled with doing this.

Reply:
Yes, this is technically valid.

If you look at the definition of `.flush`, you can see a hard limit of `1000` events in the list. This method is called for every event received, so there's not a sneaky way for it to go over that number. The list is always 1,000 elements or less.

These are objects that we're removing, so I believe the hashing is the general universal python indexing by memory address. That being the case, this should be a completely fair calculation of the upper limit time:

```python
import random
import time

alist = [object() for i in range(1000)]

def destroy_list():
  for obj in alist.copy():
    alist.remove(obj)

s = time.time(); destroy_list(); print(time.time() - s)
```

That runs in <1 ms for me. Database interactions are typically 5 ms, so I would gladly accept this upper limit, if my evaluation is correct.

I should test this again but with Django objects, just in case they override the hash method.

Reply:
Testing with saved `InventoryUpdateEvent` objects, I get 1.1 ms for the same thing. Higher, but still not a performance concern.

Reply:
As long as this method isn't somewhere in the hot loop, I guess I'm ok with it then.

Note: `list.remove` isn't going to have anything to do with hashing.  It is going to be checking for outright equality.

Reply:
Haven't been able to find reference material on that yet, but what you say suggests this should be hitting the Django model `__eq__`, which could explain the timing difference.

https://github.com/django/django/blob/78f163a4fb3937aca2e71786fbdd51a0ef39629e/django/db/models/base.py#L593-L601

From some quick poking at the subject, it sounds like the performance crunch isn't really from _searching_ for an element, but that the act of removing it requires shifting positions of all the remaining pointers. I'm sure that could be optimized, but I'm still pretty convinced that the worst-case timing is already just fine.